File: Euclid/Network/GameServer.cs
using DotNetty.Buffers;
using DotNetty.Transport.Bootstrapping;
using DotNetty.Transport.Channels;
using DotNetty.Transport.Channels.Sockets;
using Euclid.Network.Streams;
using log4net;
using System;
using System.Collections.Generic;
using System.Net;
using System.Reflection;

namespace Euclid.Network
{
    public class GameServer
    {
        #region Fields

        private static readonly ILog m_Log = LogManager.GetLogger(MethodBase.GetCurrentMethod().DeclaringType);
        private static GameServer m_GameServer;// = new GameServer();

        private MultithreadEventLoopGroup m_BossGroup;
        private MultithreadEventLoopGroup m_WorkerGroup;

        #endregion

        #region Properties

        /// <summary>
        /// Invoke the singleton instance
        /// </summary>
        public static GameServer Instance
        {
            get
            {
                return m_GameServer;
            }
        }

        /// <summary>
        /// Get the logger
        /// </summary>
        public static ILog Logger
        {
            get
            {
                return m_Log;
            }
        }

        /// <summary>
        /// Get the main server details
        /// </summary>
        public GameAddress MainServer { get; }

        /// <summary>
        /// Get the private server details
        /// </summary>
        public GameAddress PrivateServer { get; }

        /// <summary>
        /// Get the public servers details
        /// </summary>
        public List<GameAddress> PublicServers { get; }

        #endregion

        #region Constructors

        /// <summary>
        /// GameServer constructor
        /// </summary>
        public GameServer(GameAddress mainServer, GameAddress privateServer, List<GameAddress> publicServerAddresses)
        {
            this.m_BossGroup = new MultithreadEventLoopGroup(1);
            this.m_WorkerGroup = new MultithreadEventLoopGroup(10);

            this.MainServer = mainServer;
            this.PrivateServer = privateServer;
            this.PublicServers = publicServerAddresses;
        }

        #endregion

        #region Public methods

        /// <summary>
        /// Initialise the game server by given pot
        /// </summary>
        /// <param name="port">the game port</param>
        public void InitialiseServer()
        {
            try
            {
                ServerBootstrap bootstrap = new ServerBootstrap()
                    .Group(m_BossGroup, m_WorkerGroup)
                    .Channel<TcpServerSocketChannel>()
                    .ChildHandler(new GameChannelInitializer())
                    .ChildOption(ChannelOption.TcpNodelay, true)
                    .ChildOption(ChannelOption.SoKeepalive, true)
                    .ChildOption(ChannelOption.SoReuseaddr, true)
                    .ChildOption(ChannelOption.SoRcvbuf, 1024)
                    .ChildOption(ChannelOption.RcvbufAllocator, new FixedRecvByteBufAllocator(1024))
                    .ChildOption(ChannelOption.Allocator, UnpooledByteBufferAllocator.Default);

                bootstrap.BindAsync(IPAddress.Parse(this.MainServer.IpAddress), this.MainServer.Port);
                bootstrap.BindAsync(IPAddress.Parse(this.PrivateServer.IpAddress), this.PrivateServer.Port);

                foreach (var gameAddress in PublicServers)
                {
                    bootstrap.BindAsync(IPAddress.Parse(gameAddress.IpAddress), gameAddress.Port);
                }
            }
            catch (Exception e)
            {
                m_Log.Error($"Failed to setup network listener... {e}");
            }
        }

        #endregion

        #region Static methods

        public static void CreateServer(GameAddress mainServer, GameAddress privateServer, List<GameAddress> publicServerAddresses)
        {
            m_GameServer = new GameServer(mainServer, privateServer, publicServerAddresses);
        }

        #endregion
    }
}

File: euclid/src/main/java/euclid/network/GameServer.java
package euclid.network;

import euclid.network.streams.GameAddress;
import io.netty.bootstrap.ServerBootstrap;
import io.netty.channel.ChannelOption;
import io.netty.channel.EventLoopGroup;
import io.netty.channel.nio.NioEventLoopGroup;
import io.netty.channel.socket.nio.NioServerSocketChannel;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.net.InetSocketAddress;
import java.util.List;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;

public class GameServer {

    private static final Logger log = LogManager.getLogger(GameServer.class);
    private static GameServer instance;

    private final EventLoopGroup bossGroup;
    private final EventLoopGroup workerGroup;
    private final ScheduledExecutorService scheduledExecutorService;

    private final GameAddress mainServer;
    private final GameAddress privateServer;
    private final List<GameAddress> publicServers;

    public static GameServer getInstance() {
        return instance;
    }

    public static void createServer(GameAddress mainServer, GameAddress privateServer, List<GameAddress> publicServerAddresses) {
        instance = new GameServer(mainServer, privateServer, publicServerAddresses);
    }

    private GameServer(GameAddress mainServer, GameAddress privateServer, List<GameAddress> publicServerAddresses) {
        this.bossGroup = new NioEventLoopGroup(1);
        this.workerGroup = new NioEventLoopGroup(10);
        this.scheduledExecutorService = Executors.newScheduledThreadPool(4);
        this.mainServer = mainServer;
        this.privateServer = privateServer;
        this.publicServers = publicServerAddresses;
    }

    public ScheduledExecutorService getScheduledExecutorService() {
        return scheduledExecutorService;
    }

    public GameAddress getMainServer() {
        return mainServer;
    }

    public GameAddress getPrivateServer() {
        return privateServer;
    }

    public List<GameAddress> getPublicServers() {
        return publicServers;
    }

    public void initialiseServer() {
        try {
            ServerBootstrap bootstrap = new ServerBootstrap()
                    .group(bossGroup, workerGroup)
                    .channel(NioServerSocketChannel.class)
                    .childHandler(new GameChannelInitializer())
                    .childOption(ChannelOption.TCP_NODELAY, true)
                    .childOption(ChannelOption.SO_KEEPALIVE, true)
                    .childOption(ChannelOption.SO_REUSEADDR, true)
                    .childOption(ChannelOption.SO_RCVBUF, 1024);

            bootstrap.bind(new InetSocketAddress(mainServer.getIpAddress(), mainServer.getPort()));
            bootstrap.bind(new InetSocketAddress(privateServer.getIpAddress(), privateServer.getPort()));

            for (GameAddress gameAddress : publicServers) {
                bootstrap.bind(new InetSocketAddress(gameAddress.getIpAddress(), gameAddress.getPort()));
            }

            log.info("Server listening on main port {}", mainServer.getPort());
            log.info("Server listening on private port {}", privateServer.getPort());
            for (GameAddress addr : publicServers) {
                log.info("Server listening on public port {} (room {})", addr.getPort(), addr.getRoomId());
            }
        } catch (Exception e) {
            log.error("Failed to setup network listener... {}", e.getMessage());
        }
    }
}
